### DISCUSSION TITLE
MDX Blog: Migration from Version 3 to Version 5 Keep Failing

### DISCUSSION THREAD

**Discussion by MrDesjardins** [gatsbyjs/gatsby#37200] [Help]
I have been running around several build errors for the last two evenings. I have about 800 mdx files hosted without issue using Gatsby v3. 

I am running the project on a Windows machine using WSL, and I performed a few modifications, as you can see in this [Pull Request.](https://github.com/MrDesjardins/patrickdesjardins/pull/1/files)

Earlier on I was getting:

```
 ERROR #gatsby-plugin-mdx_10003  PLUGIN
gatsbjs Invalid AST. Parsed source code did not return valid output.

```
At the moment, I am getting:

```
 ERROR #gatsby-plugin-mdx_10001  PLUGIN

Failed to compile the file "/mnt/c/code/patrickdesjardins/blog/2017/typescript-vs-flow-part-2-of-3-high-level-coding.mdx". Original error message:

Unexpected character `,` (U+002C) in name, expected a name character such as letters, digits, `$`, or `_`; whitespace before attributes; or the end of the tag
```

As well as:

```
 ERROR #gatsby-plugin-mdx_10001  PLUGIN

Failed to compile the file "/mnt/c/code/patrickdesjardins/blog/2012/asp-net-mvc-with-jquery-datepicker-and-default-custom-format-date.mdx". Original error message:

Could not parse expression with acorn: Unexpected content after expression
```

and:

```
ModuleBuildError: Module build failed (from ./node_modules/gatsby/dist/utils/babel-loader.js):
  SyntaxError: /mnt/c/code/patrickdesjardins/blog/2022/docker-nodejs-frontend-backend.mdx: Invalid left-hand side in prefix operation. (1:2)
  > 1 | ---
      |   ^
    2 | title: "How to configure Docker with two containers: Frontend and Backend"
    3 | date: "2022-04-26"
    4 | categories:

  - credentials.ts:62 instantiate
    [patrickdesjardins]/[@babel]/parser/src/parse-error/credentials.ts:62:21

  - parse-error.ts:60 toParseError
    [patrickdesjardins]/[@babel]/parser/src/parse-error.ts:60:12

  - index.ts:1474 TypeScriptParserMixin.raise
    [patrickdesjardins]/[@babel]/parser/src/tokenizer/index.ts:1474:19
```

Is there something obvious I am missing? Thank you :) 

Here is the comparison with my V3 who is working well and the current changes: https://github.com/MrDesjardins/patrickdesjardins/pull/1/files

**Comment by LekoArts** [comment]
You're also using MDX 2 now which is also quite a big change. It has other requirements on the content. This here might help: https://paulie.dev/posts/2022/08/mdx-2-breaking-changes-and-gatsby-plugin-mdx-v4/

**Reply by MrDesjardins** [reply]
@LekoArts Thanks a lot. 

For some reason, and it is all my fault here, I couldn't find in Gatsby's documentation the described behavior. 

The blog you posted is great, but I still wish for a tool to migrate the MDX easily when migrating Gatsby. I understand that it is part of the journey, but that decision trickled back as an impediment to upgrade Gatsby. It is not encouraging when you have hundreds of MDX pages without assistance to migrate them. The benefits from GatsbyJS 3 to 5 are yet to be seen, I was hoping to see the hot-reload working (it does not on Windows + WSL) or a quicker build time (take about 10 minutes with ~800 MDX).

The presence in the V3 to V4 migration [documentation ](https://www.gatsbyjs.com/plugins/gatsby-plugin-mdx/?=gatsby-plugin-mdx#updating-mdx-content) as a mention about MDX, but no mention about the errors I was getting at which sidetracked me. But, that is on me. Maybe it would be beneficial for novices to Gatsby, like me, to add a section with common errors and guide them as you did in this thread, even better during compilation to see in the console what to do. But, it is just a suggestion.

For people finding that post from the errors I posted in the original post:
I installed the EsLint plugging for MDX and noticed a lot of errors. I manually changed all the tags that needed to be escaping. The list is more exhaustive than the one provided on GatsbyJS. For future reference, here is [a good one](azure-docker-container-repository-github.mdx). The EsLint has some limitations; for example, custom injection of React component will be marked as "not defined". Also, it seems to find issues with some characters in the `title`, which is fine. However, it is good for a first pass since it is a magnitude faster than running `npm run develop`

That being said, building Gatsby results to a lot of error that are non-deterministic each time I run the build.

There is sometimes:
```
ERROR #95313  HTML.COMPILATION
Building static HTML failed for path "/blog/altering-existing-schema-components-is-not-allowed/"
See our docs page for more info on this error: https://gatsby.dev/debug-html
  81 |     return context[query].data
  82 |   } else {
> 83 |     throw new Error(
     |           ^
  84 |       `The result of this StaticQuery could not be fetched.\n\n` +
  85 |         `This is likely a bug in Gatsby and if refreshing the page does not fix it, ` +
  86 |         `please open an issue in [https://github.com/gatsbyjs/gatsby/issues`](https://github.com/gatsbyjs/gatsby/issues%60)


  WebpackError: The result of this StaticQuery could not be fetched.
```
and sometimes:  

```
 ERROR #85901  GRAPHQL.UNKNOWN
There was an error in your GraphQL query:
Unknown argument "formatString" on field "MdxFrontmatter.date".
File: gatsby-node.js:39:24
  GraphQLError: Unknown argument "formatString" on field "MdxFrontmatter.date".
```
or 

```
                                                                                                                                                                                                                                                                                                                                                                                                                         
 ERROR  UNKNOWN

🚨  ERROR: Loading "createPages" query


 ERROR #11321  API.NODE.EXECUTION

"gatsby-node.js" threw an error while running the createPages lifecycle:
Cannot read properties of undefined (reading 'allMdx')
  62 |   }
  63 |
> 64 |   const posts = result.data.allMdx.edges;
     |                             ^
  65 |   const years = new Map();
  66 |   let pageCounter = 0;
  67 |   const totalPageCount = Math.ceil(posts.length / LIMIT_BLOG_COUNT);
File: gatsby-node.js:64:29
```
or else.

**Reply by MrDesjardins** [reply]
Even if there is a lot of errors with `npm run build` and when running `npm run develop` the later can have a   http://localhost:8000/___graphql  

I notice that the queries that were working don't:

```
 query {
      allMdx {
        edges {
          node {
            id
            fields {
              slug
            }
            frontmatter {
              date(formatString: "YYYY-MM-DD")
            }
            internal {
              contentFilePath
            }
          }
        }
      }
    }
```
Does not have a `formatString` at all.  Another of my query (that was working before migration):

```
 allMdx(
      sort: { frontmatter: { date: DESC } }
      limit: $limit
      skip: $skip
      filter: { frontmatter: { date: { lte: $currentDate } } }
    ) {
      nodes {
        frontmatter {
          date(formatString: "MMMM D, YYYY")
          title
          categories
        }
        id
        fields {
          slug
        }
      }
    }
  }
```
Does not have `lte`, only a few operators.

![image](https://user-images.githubusercontent.com/1014796/206810857-2d26c787-45e6-4583-ba23-6c002e0676aa.png)

I cannot find much information online. I see that people are referring having dates between quotes, which I did and do. I do not see changes in the GraphQL schema in the migration documentation.

**Reply by MrDesjardins** [reply]
I rollbacked, but I still have a PR (https://github.com/MrDesjardins/patrickdesjardins/pull/5/files) with Gatsby 3 to 5.

I'll have to revisit soon, but at the moment, the migration is a lot of work with low guidance by the tool to direct me to the right resolutions :(

**Comment by MrDesjardins** [comment]
Okay, some improvement. The problem with the GraphQL seems to sprout from a wrong date format which was a wrong key that was pressed in 1 MDX file. I am not sure why all the other MDX files are not processed and why the whole GraphQL are not generating the schema properly but for future folk you can do a search with this regex to get all MDX with wrong date format:

```
(date: )(?!("((20[012]\d|19\d\d)|(1\d|2[0123]))-((0[0-9])|(1[012]))-((0[1-9])|([12][0-9])|(3[01]))"))
```
That will catch any `.mdx` file without the format `date: "yyyy-mm-dd"`

**Comment by MrDesjardins** [comment]
I suspect some other MDX errors. 

When building, I have random pages like this one with `unknown` error, which is not helping to solve to the real issue.

```

 ERROR  UNKNOWN

Page data from page-data.json for the failed page "/blog/asp-net-mvc-project-sql-server-redis-and-managed-with-vsts/": {
  "componentChunkName": "component---src-templates-blog-article-tsx-content-file-path-mnt-c-code-patrickdesjardinsv-5-blog-2016-asp-net-mvc-project-sql-server-redis-and-managed-with-vsts-mdx",
  "path": "/blog/asp-net-mvc-project-sql-server-redis-and-managed-with-vsts/",
  "result": {
    "data": {
      "mdx": {
        "frontmatter": {
          "title": "Asp.Net MVC Project, SQL Server, Redis and managed with VSTS",
          "date": "April 4, 2016"
        }
      }
    },
    "pageContext": {
      "id": "9547bc3a-9205-5bcd-b759-88cec5cabce9",
      "totalPages": 74,
      "frontmatter": {
        "title": "Asp.Net MVC Project, SQL Server, Redis and managed with VSTS",
        "date": "2016-04-04",
        "categories": [
          "sideproject"
        ],
        "tags": [
          "c",
          "project",
          "vsts"
        ]
      }
    }
  },
  "staticQueryHashes": [
    "3159585216"
  ],
  "slicesMap": {}
}

failed Building static HTML for pages - 57.863s
```

**Comment by MrDesjardins** [comment]
I created this other regex to find where the block code were not starting a new line:

```
^(?!```).+(```)+.+$
```

I found a couple of MDX files where there was a space before the triple single quote! When I fixed them, I was able to get a full build without error. So, could it have been that the `unknown` error was "You have a white space before a triple quote in file X, Y and Z" :) I think so!

**Comment by MrDesjardins** [comment]
Locally, the build time was 913 seconds, and without the video plugin, it went down to 875 seconds. The build was not a problem with Gatsby v3 locally and on Github, but since the migration, the build has been making memory exceptions on Github action (not locally).

That being said, [Github action](https://github.com/MrDesjardins/patrickdesjardins/actions/runs/3689965769/jobs/6246483842) continues to fail:

```
info Total nodes: 5081, SitePage nodes: 826 (use --verbose for breakdown)
success Checking for changed pages - 0.000s
success onPreExtractQueries - 0.000s
success extract queries from components - 5.461s
success write out redirect data - 0.004s
success onPostBootstrap - 0.018s
info bootstrap finished - 22.390s
success write out requires - 0.147s
success Building production JavaScript and CSS bundles - 120.673s

<--- Last few GCs --->

[4151:0x5a3d840]   242548 ms: Scavenge (reduce) 1837.8 (2081.1) -> 1837.1 (2081.1) MB, 8.6 / 0.0 ms  (average mu = 0.115, current mu = 0.033) allocation failure; 
[4151:0x5a3d840]   243802 ms: Mark-sweep (reduce) 1840.5 (2084.2) -> 1839.9 (2070.7) MB, 1240.1 / 0.0 ms  (average mu = 0.097, current mu = 0.071) allocation failure; GC in old space requested


<--- JS stacktrace --->

FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
 1: 0xb6e500 node::Abort() [node]
 2: 0xa7e6[32](https://github.com/MrDesjardins/patrickdesjardins/actions/runs/3689965769/jobs/6246483842#step:8:33)  [node]
 3: 0xd47f20 v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [node]
 4: 0xd482c7 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [node]
 5: 0xf25685  [node]
 6: 0xf26588 v8::internal::Heap::RecomputeLimits(v8::internal::GarbageCollector) [node]
 7: 0xf36a93  [node]
 8: 0xf37908 v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [node]
 9: 0xf1226e v8::internal::HeapAllocator::AllocateRawWithLightRetrySlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [node]
10: 0xf13637 v8::internal::HeapAllocator::AllocateRawWithRetryOrFailSlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [node]
11: 0xef480a v8::internal::Factory::NewFillerObject(int, v8::internal::AllocationAlignment, v8::internal::AllocationType, v8::internal::AllocationOrigin) [node]
12: 0x12b7daf v8::internal::Runtime_AllocateInYoungGeneration(int, unsigned long*, v8::internal::Isolate*) [node]
13: 0x16e99f9  [node]
Aborted (core dumped)
Error: Process completed with exit code 1[34](https://github.com/MrDesjardins/patrickdesjardins/actions/runs/3689965769/jobs/6246483842#step:8:35).
```
No problem locally. To be continued.

**Comment by MrDesjardins** [comment]
At this point, I am rolling back. 

Unfortunately, there are too many hurdles to continue migrating, with very little guidance when reaching these hurdles and little documentation on optimizing the build. I would have imagined that adding a new MDX file does not require recompiling everything, including images and videos from the existing content after five iterations of Gatsby, or at least have a way to let people scale their content without reaching a wall where an 7gig of Ram machine cannot build a solution (that was building before).

In the last few days, I realized that the architecture of plugins makes Gatsby brittle by delegating some of the work (and issues) outside the core product. For example, Gatsby V3 to V4, with the change to MDX2 delegated the burden for people to update their MDX file while not providing a tool to migrate or validate. I understand that the MDX1 to MDX2 is not part of the primary mission of Gatsby, but it remains a decision Gatsby took that affect customer of the product. Similarly, plugins were still for a prior Gatsby's version for months after Gatsby updated. Similarly, the experience of having a malformed MDX generates an obscure error and little guidance. Now, there is a memory leak, and the boundary between who (plugins or Gatsby) is hard to draw, making it hard for consumers to act. Ultimately, the consumer does not necessarily want to get into the fine-grained detail and wants their content to render and have a solution that scale as they grow content.

I might be an exceptional case with the number of blog articles and media files. I have 1198 png, 30 mp4, and 736 MDX files for reference. But Gatsby V3 works with that amount of files, not Gatsby V5.

I'll remain on V3 until I find an alternative framework to host MDX files. I hope this thread helps some people fix their issues or the Gatsby team evaluate potential migration improvements for their V6. Have a great day.
